Implement module "site" for configuring distributed monitoring

lgetwan opened the pull request:
<!--- Please provide a brief summary of your changes as a title in the textbox above -->

<!---
Please use the devel branch as the merge target (see dropdown above)!
We use that branch as a staging area, to make sure the main branch
stays as stable as possible, in case people are using it to install the collection directly.
 -->

## Pull request type
<!-- Please try to limit your pull request to one type, submit multiple pull requests if needed. -->

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

There was no module for managing distributed monitoring sites.

## What is the new behavior?
<!-- Please describe the behavior or changes that are being added by this PR. -->

Now there is.

## Other information
<!-- Any other information that is important to this PR, e.g screenshots of how the component looks before and after the change. -->

Thanks to the usage of IDs in the site management REST API, it was possible to make this module idempotent. If you run into problems, please let us know.

robin-checkmk reviewed (commented):
Can we discuss the module name?
The UI calls it "Distributed monitoring", but the REST API "Site Management".
My issue here is, that the latter sounds **a lot** like what our `server` role does.
We also follow the UI in other places IIRC, as the REST API made some interesting naming choices.

lgetwan commented:
I also disliked the module name tbo, so I changed it to _distmon_, which is short and recognizable. 
Also removed Python Typing to make it Python2.7 compatible.
Changelog was also added lately.

robin-checkmk commented on plugins/doc_fragments/distmon_options.py:
```diff
@@ -0,0 +1,332 @@
+                        - Only required when the O(state) is V(login).
+                    type: dict
+                    suboptions:
+                        remote_username:
```
I think just `user` would suffice, right? We explain, that the user needs admin privileges in the description anyway. And it should be obvious, that this user has to be from the remote site.

lgetwan replied:
Changing it to "username" was easier (because then, it's the same field name as used in the REST API).

robin-checkmk commented on plugins/doc_fragments/distmon_options.py:
```diff
@@ -0,0 +1,332 @@
+                            description:
+                                - A user with administrative permissions.
+                            type: str
+                        remote_password:
```
I think just `password` would suffice, right?

lgetwan replied:
Done.

robin-checkmk commented on plugins/doc_fragments/distmon_options.py:
```diff
@@ -0,0 +1,332 @@
+                - The settings of the site.
+            type: dict
+            suboptions:
+                login_data:
```
Nitpick: Maybe `authentication`?

robin-checkmk commented on plugins/doc_fragments/distmon_options.py:
```diff
@@ -0,0 +1,332 @@
+                                        - localhost.
+                                    type: dict
+                                    suboptions:
+                                        socket_type:
```
Maybe just `type`? Looking at the structure it would be `site_config.status_connection.connection.type`. I think that is clear enough.

robin-checkmk commented on plugins/doc_fragments/distmon_options.py:
```diff
@@ -0,0 +1,332 @@
+                                                - When the connection name is unix,
+                                                - this is the path to the unix socket.
+                                            type: str
+                                proxy:
```
Maybe this should actually be `liveproxy`, just to avoid confusion?

lgetwan replied:
I will stick with the field name used by the REST API (see other comments).

robin-checkmk commented on plugins/doc_fragments/distmon_options.py:
```diff
@@ -0,0 +1,332 @@
+                                        - The Livestatus Proxy Daemon configuration attributes.
+                                    type: dict
+                                    suboptions:
+                                        use_livestatus_daemon:
```
I am not happy with this bulky option. Maybe we abstract this like `enabled` with the values `true` and `false`, resulting in `with_proxy` and `direct` respectively. Not sure if that makes enough sense, but feels cleaner.

lgetwan replied:
I will stick with the field name used by the REST API (see other comments).

robin-checkmk commented on plugins/doc_fragments/distmon_options.py:
```diff
@@ -0,0 +1,332 @@
+                                        - unreachable.
+                                    type: int
+                                    default: 2
+                                persistent_connection:
```
`persistent` should suffice.

lgetwan replied:
I will stick with the field name used by the REST API (see other comments).

robin-checkmk commented on plugins/doc_fragments/distmon_options.py:
```diff
@@ -0,0 +1,332 @@
+                                        - I(is set to) V(with_proxy)!
+                                    type: dict
+                                    suboptions:
+                                        status_host_set:
+                                            description:
+                                                - enabled for 'use the following status host' and
+                                                - disabled for 'no status host'
+                                            type: str
+                                            choices: [enabled, disabled]
+                                            default: disabled
```
Shouldn't we be able to determine this from the provided values? No values mean `disabled`, values for `host` and `site` mean, it is enabled. Or am I missing something here?

lgetwan replied:
I will stick with the way the REST API handles this (see other comments).
Unfortunately, "status_host_set" is a mandatory field, even with config replication turned off.

robin-checkmk commented on plugins/doc_fragments/distmon_options.py:
```diff
@@ -0,0 +1,332 @@
+                                            description:
+                                                - The host name of the status host.
+                                            type: str
+                                disable_in_status_gui:
```
Maybe `disable` is good enough here, as the details are in the description.

lgetwan replied:
I will stick with the field name used by the REST API (see other comments).

robin-checkmk commented on plugins/doc_fragments/distmon_options.py:
```diff
@@ -0,0 +1,332 @@
+                                - A site's configuration connection.
+                            type: dict
+                            suboptions:
+                                enable_replication:
```
Just `replication`.

lgetwan replied:
I will stick with the field name used by the REST API (see other comments).

robin-checkmk commented on plugins/doc_fragments/distmon_options.py:
```diff
@@ -0,0 +1,332 @@
+                                        - from the central sites.
+                                    type: bool
+                                    default: false
+                                url_of_remote_site:
```
Just `url`.

robin-checkmk commented on plugins/doc_fragments/distmon_options.py:
```diff
@@ -0,0 +1,332 @@
+                                        - and fix them, though!
+                                    type: bool
+                                    default: false
+                                direct_login_to_web_gui_allowed:
```
Maybe `login_to_web_gui`?

lgetwan replied:
I will stick with the field name used by the REST API (see other comments).

robin-checkmk commented on plugins/doc_fragments/distmon_options.py:
```diff
@@ -0,0 +1,332 @@
+                                                - A list of ldap connections to synchronize.
+                                            type: list
+                                            elements: str
+                                replicate_event_console:
```
Just `event_console`.

lgetwan replied:
I will stick with the field name used by the REST API (see other comments).

robin-checkmk commented on plugins/doc_fragments/distmon_options.py:
```diff
@@ -0,0 +1,332 @@
+                                        - the Event Console of the remote site.
+                                    type: bool
+                                    default: true
+                                replicate_extensions:
```
Just `extensions`.

lgetwan replied:
I will stick with the field name used by the REST API (see other comments).

robin-checkmk commented on plugins/doc_fragments/distmon_options.py:
```diff
@@ -0,0 +1,332 @@
+                                        - files below C($OMD_ROOT/local/) on the remote site will be removed.
+                                    type: bool
+                                    default: true
+                        secret:
+                            description:
+                                - The shared secret used by the central site to authenticate with the
+                                - remote site for configuring Checkmk.
```
This seems to be REST API specific, at least I cannot find it in the UI. Do we even need this here?

lgetwan replied:
When doing a GET and then using the data gathered with a POST, this field should be used. Otherwise the central site will no longer be logged in to that remote site.
Maybe, this parameter is not needed, yet. But once, we provide a lookup module, it should be there. So, I suggest to keep it.

robin-checkmk replied:
Maybe we can extend the documentation a little then? With what you just wrote it makes sense, while before I did not get the full picture.

lgetwan replied:
To be honest, I didn't find a (currently supported) use-case that included setting this secret. Thus, I removed the parameter. The module only uses the secret internally to determine if Checkmk is logged in to a remote site.

robin-checkmk commented on plugins/doc_fragments/distmon_options.py:
```diff
@@ -0,0 +1,332 @@
+                                    description:
+                                        - The customer of the site (Managed Edition - CME only).
+                                    type: str
+                                site_id:
+                                    description:
+                                        - The site ID.
+                                    type: str
```
This seems redundant as we already have a `site_id` higher up, but it is probably technically not possible to merge them, right?

robin-checkmk replied:
That makes sense. Then let's stick with approach 1 for the initial version, and we will see, what the feedback looks like. We can still do an iteration, if we learn, that the current approach does not work for users.

robin-checkmk commented on plugins/doc_fragments/distmon_options.py:
```diff
@@ -0,0 +1,332 @@
+                                        - Unfortunately, this field is required by the REST API,
+                                        - even if there is no configuration connection enabled.
+                                    type: str
+                                    default: http://localhost/nonexistant/check_mk/
```
Do we set this intentionally to an unlikely hostname, or could we also go with a nicer name?

lgetwan replied:
Actually, we should open an internal ticket to fix this behavior of the REST API. I hope, they fix this quickly, so I suggest to just keep this "nonexistant" site.

robin-checkmk replied:
"we" as in you will, or "we" as in I should do it? :wink:

lgetwan replied:
Done.

robin-checkmk reviewed (changes requested):
I am getting the overall feeling, that it might make sense, to abstract the module more from the REST API. The structure of the module could probably be flatter than that of the API. That is not to say, that this is not a good first version, we just might want to keep it in mind for a later review.

lgetwan commented:
Currently, the integration tests are not correct. I have to refine them. But from my point of view, the module itself is ready for some first manual tests.

lgetwan commented:
Looks like the integration tests runs into this problem: https://github.com/actions/runner-images/issues/6680
Probably, the tests take too long?

robin-checkmk commented:
> Looks like the integration tests runs into this problem: [actions/runner-images#6680](https://github.com/actions/runner-images/issues/6680) Probably, the tests take too long?

I'll investigate, but I doubt, it is a timeout issue, we have/had tests, which run longer. It might be resource exhaustion.

lgetwan commented:
It was indeed resource exhaustion. I reduced the number of remote sites to be created. Now github is happy.

robin-checkmk reviewed (approved):
I am happy with the changes. Pending a code review, we are good to merge.

robin-checkmk commented on changelogs/fragments/site.yml:
```diff
@@ -0,0 +1,3 @@
+major_changes:
+  - Site module - Added a module for distributed monitoring.
+    Refer to the module documentation for further details.
```
I just realized, that we should document, that this module cannot accept the remote site's livestatus certificate.
At least in the changelog under "known_issues", but probably a note in the module documentation itself might make sense.

robin-checkmk replied:
I added a hint, let me know, what you think.

lgetwan replied:
That's an important advise for the user. Good idea.

robin-checkmk commented:
The linting failures are irrelevant to this PR. I will fix them in the `devel` branch.

mike1098 commented:
I tested the following without issues against 2.2.0p33cee:

- Creating a site connection
- Loging in to a site
- Changing an attribute of a site connection
- Logout from a site
- Remove a site

Will discuss with my colleague as soon as he is back from vacation but I would say its working good.
Thank you for all the effort spent.
